Sources/NIOSOCKS/Messages/Messages.swift

```diff
@@ -0,0 +1,63 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Sent by the client and received by the server.
+public enum ClientMessage: Hashable {
+    
+    /// Contains the proposed authentication methods.
+    case greeting(ClientGreeting)
+    
+    /// Instructs the server of the target host, and the type of connection.
+    case request(SOCKSRequest)
+    
+    /// Used to respond to server authentication challenges
+    case authenticationData(ByteBuffer)
+}
+
+/// Sent by the server and received by the client.
+public enum ServerMessage: Hashable {
+    
+    /// Used by the server to instruct the client of the authentication method to use.
+    case selectedAuthenticationMethod(SelectedAuthenticationMethod)
+    
+    /// Sent by the server to inform the client that establishing the proxy to the target
+    /// host succeeded or failed.
+    case response(SOCKSResponse)
+    
+    /// Used when authenticating to send server challenges to the client.
+    case authenticationData(ByteBuffer)
+    
+    /// This is a faux message to update the server's state machine. It should be sent
+    /// once the server is satisified that the client is fully-authenticated.
+    case authenticationComplete
```

Comment:
Right now our implementation relies on the server implementer sending this essentially concurrently with their authentication data response. If for any reason it's delayed we'll have trouble, as we'll pass the SOCKS client request message through as though it were an auth message.

It might be better to have `authenticationData` carry a flag that encodes whether this is the last message in the flow. Clients can't use that, but we're doing something a bit different for clients anyway.

Reply:
Yeh you're right